src/Elmish.WPF/WpfProgram.fs

```diff
@@ -57,16 +57,36 @@ module WpfProgram =
     let bindingsLogger = program.LoggerFactory.CreateLogger("Elmish.WPF.Bindings")
     let performanceLogger = program.LoggerFactory.CreateLogger("Elmish.WPF.Performance")
 
-    let setState model dispatch =
+    (*
+     * Capture the dispatch function before wrapping it with Dispatcher.InvokeAsync
+     * so that the UI can sychrononously dispatch messages.
+     * In additional to being slightly more efficient,
+     * it also helps keep WPF in the correct state.
+     * https://github.com/elmish/Elmish.WPF/issues/371
+     * https://github.com/elmish/Elmish.WPF/issues/373
+     *
+     * This is definitely a hack.
+     * Maybe something with Elmish can change so this hack can be avoided.
+     *)
+    let mutable dispatch = Unchecked.defaultof<Dispatch<'msg>>
+
+    let setState model _ =
       match viewModel with
       | None ->
-          let vm = ViewModel<'model, 'msg>(model, dispatch, program.Bindings, program.PerformanceLogThreshold, "main", bindingsLogger, performanceLogger)
+          let uiDispatch msg = element.Dispatcher.Invoke(fun () -> dispatch msg)
+          let vm = ViewModel<'model, 'msg>(model, uiDispatch, program.Bindings, program.PerformanceLogThreshold, "main", bindingsLogger, performanceLogger)
           element.DataContext <- vm
           viewModel <- Some vm
       | Some vm ->
           vm.UpdateModel model
 
-    let uiDispatch (innerDispatch: Dispatch<'msg>) : Dispatch<'msg> =
+    let cmdDispatch (innerDispatch: Dispatch<'msg>) : Dispatch<'msg> =
+      dispatch <- innerDispatch
+      (*
+       * Have commands asychrononously dispatch messages.
+       * This avoids race conditions like can occur when shutting down.
+       * https://github.com/elmish/Elmish.WPF/issues/353
+       *)
       fun msg -> element.Dispatcher.InvokeAsync(fun () -> innerDispatch msg) |> ignore
```

Comment:
I'd revert back to Dispatcher.Invoke.

Not sure what exactly shutdown situation was like, but maybe you could check here if `Dispatcher` is disposed before calling on it? Or check inside the lambda?

Unrelated: not sure what the `InvokeAsync` mechanics are, but I've seen exceptions disappearing if you pipe a task into `ignore`... do not recommend.

Reply:
> I'd revert back to Dispatcher.Invoke.

Using `Dispatcher.Invoke` there would reintroduce issue #353.  Users can implement their own fix as I did in PR #355 to avoid that issue.  By using `Dispatcher.InvokeAsync`, we are making it easier for users.  Instructions to reproduce this issue are given in https://github.com/elmish/Elmish.WPF/pull/355#issue-581669679.

> Not sure what exactly shutdown situation was like, but maybe you could check here if `Dispatcher` is disposed before calling on it? Or check inside the lambda?

That is how I initially tried to solve this problem.   However, nothing I tried worked.  I documented everything I tried in https://github.com/elmish/Elmish.WPF/issues/330#issuecomment-773346512.

> Unrelated: not sure what the `InvokeAsync` mechanics are, but I've seen exceptions disappearing if you pipe a task into `ignore`... do not recommend.

This is a good critique.  I haven't considered what to do about that yet.  Technically the return type is [`DispatcherOperation`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcheroperation), but your point still stands.  For example, I could subscribe to the [`Aborted` event](https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcheroperation.aborted) and have it write to the logger.

Reply:
The problem with ignoring a task is not that it doesn't log an exception, but that exception handlers are not invoked at all even if you have them unless something awaits on the task.

Reply:
We could also put `try ... with e -> // log` inside the call to `InvokeAsync`.